New classical expr.Range specification

arthurostrauss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
This PR introduces a new feature enabling the specification of Range objects that can contain real-time expressions for start, stop and step. This provides a Qiskit entrypoint for OpenQASM3 RangeDefinition specification, and opens new ways of performing dynamically evolving for loops that can now depend on a wider set of expressions than the compile time specification of index set.
✅ Closes #13725 
✅ Closes #13729
### Details and comments
The Range expression is now one of the core expressions available in the expr module, as suggested in the feedback of #14277. This specification (written in Rust to comply with other expr specifications) also involves the rewriting of the OpenQASM3 exporter to include special handling of this new object, and also updates the QIskit internal OpenQASM3 AST to comply with the latest AST of the official OpenQASM3 specification (notably by specifying the type of the looping variable).

arthurostrauss committed:
Adds range expression to classical circuits

Adds a `Range` expression to the classical circuit module.

This change introduces a new `Range` expression that represents a
sequence of values, similar to Python's `range` function. This allows
for more concise and expressive representation of iterative logic within
classical circuits, particularly within `for` loops.

The `Range` expression includes start, stop, and optional step values,
and its integration involves updates to expression handling, visitor
patterns, and OpenQASM 3 export.

arthurostrauss committed:
Adds range expression to classical circuits

Adds a `Range` expression to the classical circuit module.

This change introduces a new `Range` expression that represents a
sequence of values, similar to Python's `range` function. This allows
for more concise and expressive representation of iterative logic within
classical circuits, particularly within `for` loops.

The `Range` expression includes start, stop, and optional step values,
and its integration involves updates to expression handling, visitor
patterns, and OpenQASM 3 export.

arthurostrauss committed:
Adds range expression to classical circuits

Adds a `Range` expression to the classical circuit module.

This change introduces a new `Range` expression that represents a
sequence of values, similar to Python's `range` function. This allows
for more concise and expressive representation of iterative logic within
classical circuits, particularly within `for` loops.

The `Range` expression includes start, stop, and optional step values,
and its integration involves updates to expression handling, visitor
patterns, and OpenQASM 3 export.

arthurostrauss committed:
Adds range expression to classical circuits

Adds a `Range` expression to the classical circuit module.

This change introduces a new `Range` expression that represents a
sequence of values, similar to Python's `range` function. This allows
for more concise and expressive representation of iterative logic within
classical circuits, particularly within `for` loops.

The `Range` expression includes start, stop, and optional step values,
and its integration involves updates to expression handling, visitor
patterns, and OpenQASM 3 export.

arthurostrauss committed:
Fixed bug

arthurostrauss committed:
Fixed bug

arthurostrauss committed:
Fixed bug

arthurostrauss committed:
Fixed bug

arthurostrauss committed:
Reformat

arthurostrauss committed:
Reformat

arthurostrauss committed:
Reformat

arthurostrauss committed:
Reformat

arthurostrauss committed:
Linting + bug in attribute getting in test

arthurostrauss committed:
Linting + bug in attribute getting in test

arthurostrauss committed:
Linting + bug in attribute getting in test

arthurostrauss committed:
Linting + bug in attribute getting in test

arthurostrauss committed:
Fixed additional test

arthurostrauss committed:
Fixed additional test

arthurostrauss committed:
New linting

arthurostrauss committed:
New linting

arthurostrauss committed:
New linting

arthurostrauss committed:
New linting

arthurostrauss committed:
Enables type specification for Range expressions

Adds functionality to specify the data type of Range expressions, ensuring that start, stop, and step values are appropriately cast to the specified type.

This change restricts Range values to integer types and introduces explicit casting to ensure type consistency, thereby improving type safety and preventing unexpected behavior. It also ensures that type casting errors are handled correctly.

Adds unit tests to validate the new functionality, including checks for correct type casting and error handling.

arthurostrauss committed:
Enables type specification for Range expressions

Adds functionality to specify the data type of Range expressions, ensuring that start, stop, and step values are appropriately cast to the specified type.

This change restricts Range values to integer types and introduces explicit casting to ensure type consistency, thereby improving type safety and preventing unexpected behavior. It also ensures that type casting errors are handled correctly.

Adds unit tests to validate the new functionality, including checks for correct type casting and error handling.

arthurostrauss committed:
Enables type specification for Range expressions

Adds functionality to specify the data type of Range expressions, ensuring that start, stop, and step values are appropriately cast to the specified type.

This change restricts Range values to integer types and introduces explicit casting to ensure type consistency, thereby improving type safety and preventing unexpected behavior. It also ensures that type casting errors are handled correctly.

Adds unit tests to validate the new functionality, including checks for correct type casting and error handling.

arthurostrauss committed:
Enables type specification for Range expressions

Adds functionality to specify the data type of Range expressions, ensuring that start, stop, and step values are appropriately cast to the specified type.

This change restricts Range values to integer types and introduces explicit casting to ensure type consistency, thereby improving type safety and preventing unexpected behavior. It also ensures that type casting errors are handled correctly.

Adds unit tests to validate the new functionality, including checks for correct type casting and error handling.

arthurostrauss committed:
Added  dummy __len__ method to Range + fix Cast logic

The Cast logic was refactored for the case a specific type was specified, and there is also an implicit casting done to comply with OpenQASM3 rules for promotion

arthurostrauss committed:
Added  dummy __len__ method to Range + fix Cast logic

The Cast logic was refactored for the case a specific type was specified, and there is also an implicit casting done to comply with OpenQASM3 rules for promotion

arthurostrauss committed:
Added  dummy __len__ method to Range + fix Cast logic

The Cast logic was refactored for the case a specific type was specified, and there is also an implicit casting done to comply with OpenQASM3 rules for promotion

arthurostrauss committed:
Added  dummy __len__ method to Range + fix Cast logic

The Cast logic was refactored for the case a specific type was specified, and there is also an implicit casting done to comply with OpenQASM3 rules for promotion

arthurostrauss committed:
Additional expr fix

arthurostrauss committed:
Additional expr fix

arthurostrauss committed:
Added extra check on specified Range type and corrected tests

arthurostrauss committed:
Added extra check on specified Range type and corrected tests

arthurostrauss committed:
linting

arthurostrauss committed:
linting

arthurostrauss committed:
Added new release note

arthurostrauss committed:
Added new release note

arthurostrauss committed:
Adds support for range expressions

Extends the expression system to include range expressions, allowing for more flexible and expressive manipulation of data within the circuit.

This change introduces `Expr::Range` and handles it in the expression reference, mutable expression reference, and visitor implementations. It also adds a comparison check for range expressions to ensure they have the same type and constant property.

arthurostrauss committed:
Adds support for range expressions

Extends the expression system to include range expressions, allowing for more flexible and expressive manipulation of data within the circuit.

This change introduces `Expr::Range` and handles it in the expression reference, mutable expression reference, and visitor implementations. It also adds a comparison check for range expressions to ensure they have the same type and constant property.

arthurostrauss committed:
Linting fix attempt

arthurostrauss committed:
Linting fix attempt

arthurostrauss commented:
@1ucian0, the current tests that are failing have a priori no connection with the implemented features of this PR: 
2 tests, related to high level synthesis and layoutTransformation. In both cases, the error seems to come from two circuits that do not look equivalent. Would it be possible that the seeding might have changed the way the circuit is transformed ? 
I can't see why my feature would make those tests fail. Any assistance would be appreciated.

1ucian0 commented on test/python/circuit/classical/test_range.py:
```diff
@@ -0,0 +1,208 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
```
This is a new file
```suggestion
# (C) Copyright IBM 2025.
```

1ucian0 commented on releasenotes/notes/expr_range_feature-1091deab8f458eae.yaml:
The exporter now is explicit about the loop variable type, which correctly follows the standard   https://github.com/openqasm/openqasm/pull/521 I think an upgrade release note could help users to be aware of the change.

1ucian0 commented on qiskit/circuit/controlflow/for_loop.py:
```diff
@@ -178,7 +180,7 @@ def __init__(
         # We can pass through `range` instances because OpenQASM 3 has native support for this type
         # of iterator set.
-        self._indexset = indexset if isinstance(indexset, range) else tuple(indexset)
+        self._indexset = indexset if isinstance(indexset, (range, Range)) else tuple(indexset)
```
Do we expect users to directly use `expr.Range` in `ForLoopOp`? If yes, I think it should be tested. If not, I think the extensions on in this file can be removed.

1ucian0 commented on qiskit/circuit/quantumcircuit.py:
Ditto previous comment about the direct use of `expr.Range` in `ForLoopOp`

1ucian0 commented on qiskit/qasm3/ast.py:
```diff
@@ -679,17 +679,19 @@ class ForLoopStatement(Statement):
         indexset: Union[Identifier, IndexSet, Range],
         parameter: Identifier,
         body: ProgramBlock,
+        type: ClassicalType,
```
while `qasm3.ast.ForLoopStatement` is not part of the public API, I think it would be better to kvarg `type=None` and raise a deprecation warning when not set.

1ucian0 replied:
it is a good practice to avoid build-in name shadowing.
```suggestion
        type_: ClassicalType,
```

1ucian0 commented on test/python/circuit/classical/test_range.py:
```diff
@@ -0,0 +1,208 @@
+from test import QiskitTestCase  # pylint: disable=wrong-import-order
+
+
+@ddt.ddt
```
I dont think you are using ddt in these tests.

1ucian0 commented on test/python/circuit/classical/test_range.py:
```diff
@@ -0,0 +1,208 @@
+
+    def test_range_with_non_constant_values(self):
+        """Test creating a Range with non-constant values."""
+        from qiskit.circuit import ClassicalRegister
```
a detail, we usually import `ClassicalRegister` from root
```suggestion
        from qiskit import ClassicalRegister
```

1ucian0 reviewed (commented):
Thanks Arthur! I took a look to the python part and added few comments.

arthurostrauss committed:
Added default step value to 1 in promoted type

Also added tests for the for loop usage with the range expression.
Updated the release notes to account for the change of the OpenQASM3 exporter

arthurostrauss committed:
Added default step value to 1 in promoted type

Also added tests for the for loop usage with the range expression.
Updated the release notes to account for the change of the OpenQASM3 exporter

arthurostrauss committed:
Reformat and small fix on arg name

arthurostrauss committed:
Reformat and small fix on arg name

arthurostrauss committed:
Linting

arthurostrauss committed:
Linting

arthurostrauss committed:
Fixed tests and linting

arthurostrauss committed:
Fixed tests and linting

arthurostrauss committed:
Additional linting

arthurostrauss committed:
Additional linting

arthurostrauss committed:
Update release note section name

arthurostrauss committed:
Update release note section name

arthurostrauss committed:
Release notes rewriting

arthurostrauss committed:
Release notes rewriting

arthurostrauss committed:
additional editing of release note

arthurostrauss committed:
additional editing of release note

arthurostrauss commented:
@1ucian0 

Still waiting for another review round on this PR.

Thank you.

arthurostrauss committed:
Linting fix

arthurostrauss committed:
Linting fix

arthurostrauss commented on crates/circuit/src/classical/expr/range.rs:
```diff
@@ -0,0 +1,331 @@
+    module = "qiskit._accelerate.circuit.classical.expr"
+)]
+#[derive(PartialEq, Clone, Debug)]
+pub struct PyRange(pub Range);
```
Any suggestion for another name?

arthurostrauss committed:
Refactors range expressions to ensure step is always present

This change modifies the range expression representation.

It ensures that the `step` within a range is always an `Expr`, defaulting to 1 when not explicitly provided, rather than an `Option`.
This simplifies subsequent logic by removing the need for `.unwrap()` calls when accessing the step value.

It also adds structural equivalence checks for start, stop, and step expressions within range expressions for more accurate comparison.

arthurostrauss committed:
Refactors range expressions to ensure step is always present

This change modifies the range expression representation.

It ensures that the `step` within a range is always an `Expr`, defaulting to 1 when not explicitly provided, rather than an `Option`.
This simplifies subsequent logic by removing the need for `.unwrap()` calls when accessing the step value.

It also adds structural equivalence checks for start, stop, and step expressions within range expressions for more accurate comparison.

arthurostrauss committed:
Reformat

arthurostrauss committed:
Reformat

arthurostrauss committed:
Linting

arthurostrauss committed:
Linting

arthurostrauss committed:
Updates range representation in Python

Modifies the `__repr__` method of the `PyRangeExpr` class to use a more concise and readable string format for representing range expressions in Python.
It now displays the range as "Range(start, stop, step, ty)" instead of the previous "R(start=..., stop=..., step=..., ty=...)".

arthurostrauss committed:
Updates range representation in Python

Modifies the `__repr__` method of the `PyRangeExpr` class to use a more concise and readable string format for representing range expressions in Python.
It now displays the range as "Range(start, stop, step, ty)" instead of the previous "R(start=..., stop=..., step=..., ty=...)".

arthurostrauss committed:
Removes unnecessary formatting in repr

Removes redundant formatting within the repr method for enhanced code readability.

arthurostrauss committed:
Removes unnecessary formatting in repr

Removes redundant formatting within the repr method for enhanced code readability.

gadial reviewed (dismissed):
The changes look good to me, I approve the part I've looked at (the rust code)

1ucian0 commented on crates/circuit/src/classical/expr/range.rs:
```diff
@@ -69,7 +72,11 @@ fn py_value_to_expr(_py: Python, value: &Bound<PyAny>) -> PyResult<Expr> {
+                .get_type()
+                .name()
+                .map(|s| s.to_string())
+                .unwrap_or_else(|_| "<missing name>".to_string())
```
It’s fine if you use AI-assisted coding. However, it’s expected that you can explain and support the decisions in the code you’re submitting.

arthurostrauss replied:
I agree, however I only got very recently acquainted with Rust and am yet shaky on the interfaces with Python, so this was the fix suggested as the default string yielded an error.

gadial replied:
I actually wonder about two other things:
1. The `to_string` used in `"<missing name>".to_string()` - that's the one I asked about.
2. Using `map` - why is this required?

arthurostrauss committed:
Removed unnecessary logic on error message

In the error message, we just want to indicate the type of the variable and not its value (nor its name) so we can simplify the error message

arthurostrauss committed:
Removed unnecessary logic on error message

In the error message, we just want to indicate the type of the variable and not its value (nor its name) so we can simplify the error message

arthurostrauss committed:
Linting

arthurostrauss committed:
Linting

arthurostrauss committed:
Update range.rs (linting)

arthurostrauss committed:
Update range.rs (linting)

1ucian0 commented:
I think there are still some unaddressed comments from @gadial

arthurostrauss committed:
Update qiskit/qasm3/ast.py

Co-authored-by: Luciano Bello <766693+1ucian0@users.noreply.github.com>

arthurostrauss committed:
Update qiskit/qasm3/ast.py

Co-authored-by: Luciano Bello <766693+1ucian0@users.noreply.github.com>

arthurostrauss committed:
Update releasenotes

Co-authored-by: Luciano Bello <766693+1ucian0@users.noreply.github.com>

arthurostrauss committed:
Update releasenotes

Co-authored-by: Luciano Bello <766693+1ucian0@users.noreply.github.com>

arthurostrauss committed:
Update releasenotes

Co-authored-by: Luciano Bello <766693+1ucian0@users.noreply.github.com>

arthurostrauss committed:
Update releasenotes

Co-authored-by: Luciano Bello <766693+1ucian0@users.noreply.github.com>

arthurostrauss committed:
Update releasenotes

Co-authored-by: Luciano Bello <766693+1ucian0@users.noreply.github.com>

arthurostrauss committed:
Update releasenotes

Co-authored-by: Luciano Bello <766693+1ucian0@users.noreply.github.com>

arthurostrauss commented on releasenotes/notes/expr_range_feature-1091deab8f458eae.yaml:
Can you clarify what you mean by this? The deprecation should be indicated in a separate section you mean?

1ucian0 replied:
In the `upgrade` sections of the release notes, you can warning users about API changes that are coming that, eventually, will be breaking changes. You are introducing a deprecation warning for typeless for-loops. I think you should mention it in one of those `upgrade` sections.

arthurostrauss replied:
A note was added accordingly.

arthurostrauss commented:
> I think there are still some unaddressed comments from @gadial

I resolved the conversations that were left unresolved, but were addressed in my previous changes. @gadial Feel free to add anything if needed.

arthurostrauss committed:
Linting

arthurostrauss committed:
Linting

arthurostrauss committed:
Added copy and pickle method to comply with transpiler constraints

Copy and pickle methods were implemented to enable the usage of advanced optimization levels in transpiler on circuits containing Range objects

arthurostrauss committed:
Added copy and pickle method to comply with transpiler constraints

Copy and pickle methods were implemented to enable the usage of advanced optimization levels in transpiler on circuits containing Range objects

arthurostrauss committed:
Reformat

arthurostrauss committed:
Reformat

arthurostrauss committed:
Updated upgrade section of release notes

arthurostrauss committed:
Updated upgrade section of release notes

arthurostrauss commented on test/python/circuit/classical/test_range.py:
```diff
@@ -0,0 +1,285 @@
+        self.assertEqual(range_expr.type, types.Uint(8))
+        self.assertTrue(range_expr.const)
+
+    def test_range_with_float_values(self):
```
I don't think start should be the only one tested, because we want to test the type casting of the step when not explicitly provided.

arthurostrauss replied:
@1ucian0, can you indicate if the recent changes have solved the point you raised?

arthurostrauss committed:
Merge remote-tracking branch 'origin/dynamic_range_expr' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Updated Range to comply with latest Rust changes

arthurostrauss committed:
Linting

arthurostrauss commented:
@1ucian0  
Still waiting for a final approval here.

arthurostrauss committed:
Update range.rs

Additional linting

1ucian0 committed:
Merge branch 'main' into dynamic_range_expr

arthurostrauss committed:
Merge branch 'main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Adapted Range to comply with latest Value specification

arthurostrauss committed:
Upgraded test Range using ddt

arthurostrauss committed:
Fixed default type in QASM3 AST for ForLoop statement

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Adds support for range indexsets in for-loops

Extends the circuit representation to support range objects (both `expr.Range` and Python `range`) as index sets for for-loops, in addition to lists of integers.

This change ensures that for-loops can iterate over dynamically-sized ranges defined by `expr.Range`, preserving the Python `range` object when provided as input. This allows users to use Python ranges as loop iterators instead of just lists of indices.
The length of `expr.Range` objects can only be determined at runtime, a fallback mechanism is added to handle this case gracefully.

Also introduces a dedicated enum `ForLoopIndexSet` to encapsulate the different types of index sets.

arthurostrauss committed:
Improves ForLoopOp drawing with dynamic range

Enhances the visualization of `ForLoopOp` nodes in the circuit drawer.

It now correctly displays the range parameters (start, stop, step) when the index set is an `expr.Range` object, extracting and displaying their values or names.

This provides a clearer and more informative representation of loop indices in circuit diagrams, especially when using dynamic ranges.

arthurostrauss committed:
Simplifies range formatting in circuit visualization

Streamlines the formatting of range expressions displayed in circuit visualizations, improving readability.

Removes redundant conditional logic and simplifies the construction of the range string.

arthurostrauss committed:
Added is_empty() method for ForLoopIndexSet

arthurostrauss committed:
Improves display of range objects in circuit visualization

Refactors the logic for displaying range objects within circuit visualizations.

Simplifies the handling of `range` and `IndexSet` objects to provide a more consistent and readable representation, especially when dealing with step values.

The changes enhance the display of dynamic ranges to avoid fallback to string representations of the `IndexSet`.

1ucian0 committed:
Merge branch 'main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

Updated ForLoopIndexSet to match latest ForCollection definition

arthurostrauss committed:
Reduces cloning in range expression creation.

Addresses comments to reduce unnecessary cloning during the creation of range expressions.

This change reduces clone statements in the core range.rs logic. It also introduces getter methods that are now used in the `__repr__` method to further reduce unnecessary cloning.

arthurostrauss commented:
This PR has been updated with the latest modifications brought by #15404  and complies with the recent Rust porting done by #14568. It also addressed the remaining comments of @gadial on implementation details of the new range.rs file. All tests are passing and build on top of the previous changes of control-flow that were planned for Qiskit 2.3. 

@1ucian0, @jakelishman, it is ready for a last review round and could be reintegrated to the 2.3 milestone unless there are critical reasons not to do it (e.g. anything else on control-flow revamping that needs to be prioritized over adding this).

To comply with the current model of ForCollection, new methods called is_empty() and len() have been implemented in the Range object, and return by default False and 1 as both those methods would not be relevant at compile time.

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
feat: Add QPY support for `expr.Range` objects

Enables the serialization and deserialization of `ForLoopOp` instances that use dynamic ranges defined by `qiskit.circuit.classical.expr.Range`. Previously, only static Python `range` objects were supported as iterables in serialized `for` loops.

This introduces a new `RANGE` type key to the QPY expression format. The serialization logic in both the Python and Rust backends is updated to correctly handle writing and reading `Range` expression nodes, recursively handling their `start`, `stop`, and `step` components as expressions.

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Linting

arthurostrauss committed:
Update value.rs (linting)

arthurostrauss committed:
Additional linting

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
updated licensing

arthurostrauss committed:
Updated license header on remaining file

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Fix merging artifact

reformat

arthurostrauss committed:
Removed optional Type for AST statement and updated release note

arthurostrauss committed:
Reformatting

arthurostrauss committed:
Removed former type hinting

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Additional linting

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Edited qpy update

arthurostrauss committed:
Linting

arthurostrauss committed:
Removed equivalence check for subnodes in structural equivalence check

Following @gadial 's comment, removed unnecessary additional check for the Range object.

arthurostrauss commented on crates/circuit/src/classical/expr/expr.rs:
```diff
@@ -237,6 +249,11 @@ impl Expr {
                             return false;
                         }
                     }
+                    (ExprRef::Range(a), ExprRef::Range(b)) => {
```
I had overlooked this comment earlier. Indeed, based on the iterator done at the beginning of this check, there's no need to actually check all values. I removed this extra check and left it to be checking type and constant as other expressions.

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
feat: support classical expr.Range as indexset in UnrollForLoops pass

Key API changes include:
- Added a `values()` method to `expr.Range` to materialize constant bounds into a Python `range`.
- `ForLoopOp` construction now raises a `CircuitError` if a non-constant `expr.Range` is paired with a compile-time `loop_parameter` that is used in the loop body.
- `UnrollForLoops` can now unroll loops with constant `expr.Range` indexsets. A new `strict` parameter (default `False`) allows raising a `TranspilerError` when encountering non-constant ranges instead of skipping them.

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss commented:
Hi @gadial ,

I have just pushed some changes that aim at addressing the two points you raised.

1. UnrollForLoops assumed the indexset was iterable. expr.Range is not (by design), so even a fully constant range raised TypeError: 'Range' object is not iterable.

Fix: I added expr.Range.values() in Rust, which materializes a constant range to a Python range (exclusive stop, same semantics as the expression). The pass calls that before unrolling, so your example works the same as range(0, 5). Note that this method raises if the Range is not constant, e.g., contains a real-time Var that is not of type Value.

Non-constant ranges (e.g. bounds from add_var) are not unrollable at transpile time. The pass skips them by default (strict=False). With strict=True, it raises TranspilerError instead of failing silently or with a confusing error.

2. Historically, the for-loop index is a compile-time Parameter, bound when the loop is unrolled via assign_parameters. That still applies when the indexset is a constant expr.Range (or Python range / tuple).

A non-constant expr.Range means indices are only known at runtime. Using a Parameter in the body in that case would not be bound correctly, so we now raise CircuitError at construction if:

indexset is a non-constant expr.Range, and
loop_parameter is set and appears in the loop body
(including with qc.for_loop(dynamic_range) as i: qc.rx(i, 0)).

Parameterless loops (or unused auto-generated parameters that get stripped) remain valid for runtime / OpenQASM 3 export. Real-time Var as the loop index is explicitly out of scope for this PR and left for future work (this would require a significant revamping of the ForLoopOp itself, for which I would cast into some exclusive real time control flow and remove this compile time Parameter support which does not align with the other control flow operations that do exist in the same module).

3. I added more tests to check both behaviors accordingly.

arthurostrauss committed:
refactor: extract pure Rust bounds materialization for expr.Range

Introduced a `MaterializeError` enum and an internal `try_materialize_bounds` method on `Range` to extract constant integer bounds `(start, stop, step)` without requiring Python interaction. The existing `values` method now wraps this pure Rust implementation to produce the `PyRange` object, and related tests have been updated to run without a Python environment.

gadial commented:
A for a `Var` index, one reason I'm asking is because it might be helpful in order to tackle this challenge: https://github.com/Qiskit/qiskit/issues/15986

However, it's perfectly acceptable if you think it's out of scope for this PR.

arthurostrauss committed:
feat: support real-time expr.Var as loop variables in ForLoopOp

Key API changes include:
- `ForLoopOp` and `QuantumCircuit.for_loop` now support using a real-time `expr.Var` as the loop index when paired with a classical `expr.Range` indexset. Auto-generated loop variables that are unused in the body are automatically dropped.
- Added `substitute_var_in_expr` and `substitute_var_in_circuit` utilities in `qiskit.circuit.classical.expr` to recursively replace classical variables across expression trees and full circuits.
- `UnrollForLoops` now unrolls constant `expr.Range` indexsets paired with an `expr.Var` by substituting the iteration value throughout the loop body.
- Bumped QPY serialization format to version 17 to support a new `LOOP_VARIABLE` instruction parameter type for runtime loop indices.
- **Breaking Change**: Constructing a `ForLoopOp` with mismatched `indexset` and `loop_parameter` types (e.g., Python `range` with `expr.Var`, or `expr.Range` with `Parameter`) now raises a `CircuitError`.

arthurostrauss committed:
fix: preserve Parameter identity when substituting in ForLoopOp body

When substituting classical variables inside a `ForLoopOp`, if the loop body changes but the indexset does not, the operation is now updated using `replace_blocks` instead of creating a new `ForLoopOp` instance. This ensures that compile-time `Parameter` instances used as loop variables maintain their object identity, respecting `Parameter.__deepcopy__` invariants.

arthurostrauss committed:
refactor: drop Parameter identity preservation in ForLoopOp substitution

Key API changes include:
- Compile-time `Parameter` loop variables in `ForLoopOp` are now preserved by name only during classical variable substitution. Object identity with the caller's original instance is no longer guaranteed, aligning with `QuantumCircuit.append` deepcopy behavior for parameterized operations.
- Removed the `replace_blocks` workaround in `qiskit.circuit.classical.expr.substitute` that previously enforced `Parameter` identity.
- Updated internal comments in `QuantumCircuit` scope validation to clarify that `params[0]` is validated when it is an `expr.Range`, and `params[1]` may be a real-time `expr.Var` declared inside the loop body.

1ucian0 commented on qiskit/circuit/classical/expr/substitute.py:
```diff
@@ -0,0 +1,193 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2025.
```
```suggestion
# (C) Copyright IBM 2026.
```

1ucian0 commented on releasenotes/notes/expr_range_feature-1091deab8f458eae.yaml:
```diff
@@ -0,0 +1,132 @@
+
+features_transpiler:
+  - |
+    :class:`~qiskit.transpiler.passes.UnrollForLoops` now supports
```
```suggestion
    The transpiler pass :class:`~qiskit.transpiler.passes.UnrollForLoops` now supports
```

1ucian0 commented on test/python/circuit/classical/test_range.py:
```diff
@@ -0,0 +1,420 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2025.
```
```suggestion
# (C) Copyright IBM 2026.
```

1ucian0 commented on qiskit/circuit/classical/expr/substitute.py:
```diff
@@ -0,0 +1,193 @@
+    return node
+
+
+def substitute_var_in_circuit(
```
I wonder if this function could be a  method in `QuantumCircuit`. Also, why not in-place substitution? rebuilding the circuit might have performance implications.

1ucian0 commented on qiskit/circuit/classical/expr/substitute.py:
```diff
@@ -0,0 +1,193 @@
+    return _substitute_circuit(circuit, substitutions)
+
+
+def _substitute_circuit(circuit: QuantumCircuit, subs: dict) -> QuantumCircuit:
```
`substitute_var_in_circuit` seems like a direct alias of `_substitute_circuit`.

```suggestion
```

1ucian0 commented on qiskit/circuit/controlflow/builder.py:
```diff
@@ -539,6 +565,14 @@ def use_var(self, var: expr.Var):
+                self._vars_local[var.name] = var
+                del self._vars_pending[var.name]
+                return
+            raise CircuitError(f"cannot use '{var}' which is shadowed by the local '{pending}'")
```
i was unable to find a test raising this exception

1ucian0 commented on qiskit/circuit/controlflow/builder.py:
```diff
@@ -497,6 +502,25 @@ def add_uninitialized_var(self, var: expr.Var):
+        is in scope for resolution, but only emitted into the built body if actually referenced
+        in a classical expression. Used by :class:`ForLoopContext` when auto-generating a loop
+        :class:`~.expr.Var` so that an unused auto-generated variable is dropped, matching the
+        existing behavior for auto-generated :class:`~.Parameter` loop variables.
```
This docstring is not getting published (as it is not part of the public API) so no need to cross-link and it can be easier to read.

```suggestion
        This mirrors how `_vars_capture` is populated lazily by `use_var` the variable
        is in scope for resolution, but only emitted into the built body if actually referenced
        in a classical expression. Used by `ForLoopContext` when auto-generating a loop
        `~.expr.Var so that an unused auto-generated variable is dropped, matching the
        existing behavior for auto-generated :class:`~.Parameter` loop variables.
```

1ucian0 commented on qiskit/circuit/controlflow/for_loop.py:
```diff
@@ -21,34 +21,112 @@
+        )
+
+
+def _is_loop_variable_referenced(loop_parameter, body) -> bool:
```
a private method for `ForLoopContext` ?

1ucian0 commented on qiskit/circuit/controlflow/for_loop.py:
```diff
@@ -21,34 +21,112 @@
+    return loop_parameter in body.parameters
+
+
+def _register_loop_variable_in_scope(loop_parameter, scope, *, lazy: bool) -> None:
```
a private method for `ForLoopContext` ?

arthurostrauss replied:
Done in latest changes

1ucian0 commented on test/python/circuit/classical/test_substitute.py:
```diff
@@ -0,0 +1,334 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2025.
```
```suggestion
# (C) Copyright IBM 2026.
```

1ucian0 commented on crates/circuit/src/classical/expr/range.rs:
```diff
@@ -0,0 +1,600 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2025
```
```suggestion
// (C) Copyright IBM 2026
```

1ucian0 committed:
Copyright IBM 2026

1ucian0 commented on qiskit/circuit/classical/expr/substitute.py:
```diff
@@ -0,0 +1,193 @@
+    from qiskit.circuit import QuantumCircuit
+
+
+def substitute_var_in_expr(
```
Could it be an `Expr` method?

arthurostrauss replied:
It is now an Expr method in Rust

1ucian0 commented on qiskit/circuit/classical/expr/substitute.py:
```diff
@@ -0,0 +1,193 @@
+
+
+def _substitute(node: _expr.Expr, subs: dict) -> _expr.Expr:
+    if isinstance(node, _expr.Var):
```
All these cases seem like implementations in each class of `substitute_var_in_expr`

arthurostrauss replied:
Removed by porting substitution to Rust directly.

1ucian0 commented on qiskit/circuit/classical/expr/substitute.py:
```diff
@@ -0,0 +1,193 @@
+
+    for inst in circuit.data:
+        op = inst.operation
+        if isinstance(op, Store):
```
consider implementing `substitute` in each `op` class instead of this huge switch-case . Similar to https://github.com/Qiskit/qiskit/pull/14618/files#r3317259128

arthurostrauss replied:
Done in latest changes, please review

arthurostrauss committed:
Update release note

Co-authored-by: Luciano Bello <766693+1ucian0@users.noreply.github.com>

arthurostrauss committed:
Update builder.py

Co-authored-by: Luciano Bello <766693+1ucian0@users.noreply.github.com>

arthurostrauss committed:
Update range.rs

Co-authored-by: Luciano Bello <766693+1ucian0@users.noreply.github.com>

1ucian0 commented on qiskit/qpy/__init__.py:
I think this needs a bump in the QPY version

@mtreinish ?

1ucian0 replied:
Please, bump the QPY version to 18.

arthurostrauss replied:
Done

arthurostrauss committed:
refactor: move classical var substitution to Expr and QuantumCircuit methods

Key API changes include:
- Replaced the standalone functions `substitute_var_in_expr` and `substitute_var_in_circuit` with object-oriented methods: `Expr.substitute` and `QuantumCircuit.substitute_vars`.
- The new `QuantumCircuit.substitute_vars` method adds an `inplace` keyword argument (mirroring `assign_parameters`), along with `strict` and `keep_substituted_vars` arguments for finer control over variable scope and retention.
- Added a `substitute` method to operations that contain classical expressions (`Store`, `BoxOp`, `ForLoopOp`, `IfElseOp`, `SwitchCaseOp`, `WhileLoopOp`) which `QuantumCircuit` dispatches to.
- The core expression tree substitution logic is now implemented directly in Rust on the `Expr` struct.
- The standalone `substitute_var_in_expr` and `substitute_var_in_circuit` utilities have been removed from `qiskit.circuit.classical.expr`.

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Linting

arthurostrauss committed:
Formatting

arthurostrauss committed:
feat: bump QPY version to 18 for expr.Range and ForLoopOp loop variables

Key API changes include:
- Bumped the maximum QPY format version to 18 (`QPY_VERSION = 18`).
- Gated the serialization and deserialization of classical `expr.Range` expressions and `ForLoopOp` runtime loop variables behind QPY version 18 to prevent breaking compatibility with version 17 parsers.
- `qiskit.qpy.dump` now explicitly raises `UnsupportedFeatureForVersion` when attempting to serialize circuits containing `expr.Range` or `LOOP_VARIABLE` payloads with target versions below 18 (previously `LOOP_VARIABLE` was incorrectly targeted at v17).
- For QPY target versions below 18, the loop-variable slot will be emitted as `Null` (lossy by design) provided the indexset itself does not require v18.

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
test: fix QPY roundtrip test checking original circuit instead of loaded one

The test for roundtripping a `for_loop` with a loop variable mistakenly extracted the operation from the original circuit (`qc.data`) instead of the deserialized payload. This commit fixes the test logic to properly serialize and deserialize the circuit, extract the operation from the newly loaded circuit data, and updates the validation from `assertIs` to `assertEqual` to account for the deserialized object copy.

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Linting

1ucian0 commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -2894,11 +2894,18 @@ def append(
+        # For a :class:`.ForLoopOp`, ``params[1]`` may be a real-time :class:`~.expr.Var`
+        # declared inside the loop body.  It must not be validated against the outer
+        # scope.  ``params[0]`` (the indexset) is still validated when it is an
+        # :class:`~.expr.Range`.
```
comments are not parsed by sphinx
```suggestion
        # For a `ForLoopOp` class, `params[1]` may be a real-time `expr.Var`
        # declared inside the loop body.  It must not be validated against the outer
        # scope.  `params[0]` (the indexset) is still validated when it is an
        # `expr.Range` class.
```

1ucian0 commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -5073,6 +5080,249 @@ def assign_parameters(
+        strict: bool = True,
+        keep_substituted_vars: bool = False,
+    ) -> QuantumCircuit | None:
+        """Substitute real-time classical :class:`~.expr.Var` nodes throughout the circuit.
```
This is great documentation. However, it does not get compiled in the doc for some reason. Take a look to `tox -e docs`

1ucian0 committed:
Merge branch 'main' into dynamic_range_expr

1ucian0 commented on releasenotes/notes/expr_range_feature-1091deab8f458eae.yaml:
```diff
@@ -0,0 +1,156 @@
+    expression trees:
+
+    * :meth:`~qiskit.circuit.classical.expr.Expr.substitute` rewrites a single
+      :class:`~qiskit.circuit.classical.expr.Expr` node, replacing every
```
This link is broken.
```
:class:`~qiskit.circuit.classical.expr.Expr`
```

I think the docstring for the class `Range` class is not getting indexed.

arthurostrauss replied:
Indeed, should be fixed now.

1ucian0 commented on qiskit/transpiler/passes/utils/unroll_forloops.py:
```diff
@@ -57,18 +91,51 @@ def run(self, dag):
         return dag
 
 
+def _bind_loop_param(body, loop_param, index_value, var_type):
```
another loose function. It is used just once, maybe it can be inlined.

arthurostrauss committed:
feat: support expr.Range docs and inlining for unrolling ForLoopOp

Add documentation for the expression Range (loop bounds) and expose Expr members used for substitution/visitor patterns.

arthurostrauss committed:
Added referenced docs for QuantumCircuit.substitute_vars

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

1ucian0 commented on qiskit/circuit/controlflow/for_loop.py:
```diff
@@ -16,39 +16,83 @@
     from qiskit.circuit import QuantumCircuit
 
 
+def _validate_for_loop_params(indexset, loop_parameter, body) -> None:
```
another loose func that culd be inlined.

arthurostrauss replied:
Done, thanks.

arthurostrauss committed:
refactor: inline for-loop indexset/loop_parameter validation into ForLoopOp

Move the validation logic for indexset vs loop_parameter into the ForLoopOp constructor and remove the now-unreferenced helper function (_validate_for_loop_params). Behavior is unchanged: compile-time Parameters cannot be used with expr.Range indexsets and expr.Var cannot be used with Python range/list indexsets. This is an internal refactor only and does not change the public API or semantics.

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

1ucian0 commented on qiskit/circuit/controlflow/switch_case.py:
```diff
@@ -187,6 +187,32 @@ def replace_blocks(self, blocks: Iterable[QuantumCircuit]) -> SwitchCaseOp:
+    def substitute(self, substitutions: Mapping[expr.Var, expr.Expr]) -> SwitchCaseOp:
+        """Return a new :class:`SwitchCaseOp` with classical :class:`~.expr.Var` nodes replaced.
+
+        The substitution is applied to the :attr:`target` (when it is an
```
There is no documentation for :attr:`target` in stubs/qiskit.circuit.SwitchCaseOp.html, so link does not work

1ucian0 commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -5073,6 +5084,249 @@ def assign_parameters(
+            variable is not referenced anywhere.  Use this flag only when a later pass will either
+            re-substitute or explicitly remove those retained declarations.
+
+            **``strict=False`` interaction.**  When a key is *not* declared in this circuit
```
It renders like
<img width="663" height="146" alt="Screenshot 2026-06-10 at 13 22 27" src="https://github.com/user-attachments/assets/62cfcfdd-ec89-42bc-b34b-db4ed1168c06" />

1ucian0 commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -5073,6 +5084,249 @@ def assign_parameters(
+        Args:
+            substitutions: mapping from each :class:`~.expr.Var` to replace to its replacement
+                :class:`~.expr.Expr`.  For a safe, type-consistent rewrite the replacement
+                should have the same :attr:`~.expr.Expr.type` as the key it replaces.
```
broken link
```
:attr:`~.expr.Expr.type`
```

1ucian0 commented on qiskit/qpy/__init__.py:
```diff
@@ -509,6 +512,37 @@ def open(*args):
+.. _qpy_version_18:
+
+Version 18
+----------
```
Bring over the `Version 17` header (to a lower line number), so keeps the order.

1ucian0 commented on qiskit/transpiler/passes/utils/unroll_forloops.py:
```diff
@@ -12,29 +12,53 @@
+
+    For constant :class:`~.expr.Range` indexsets paired with an :class:`~.expr.Var`
+    loop variable, each iteration substitutes the :class:`~.expr.Var` with the
+    iteration value (an :class:`~.expr.Value`) in every classical expression in
```
```suggestion
    iteration value (of type :class:`~.expr.Value`) in every classical expression in
```

1ucian0 commented on releasenotes/notes/expr_range_feature-1091deab8f458eae.yaml:
```diff
@@ -0,0 +1,156 @@
+    The transpiler pass :class:`~qiskit.transpiler.passes.UnrollForLoops` now supports
+    :class:`~qiskit.circuit.ForLoopOp` instructions whose ``indexset`` is a
+    **constant** :class:`~qiskit.circuit.classical.expr.Range`. The pass calls
+    :meth:`~qiskit.circuit.classical.expr.Range.values` to materialize the bounds
```
```suggestion
    :meth:`expr.Range.values` to materialize the bounds
```

1ucian0 commented on releasenotes/notes/expr_range_feature-1091deab8f458eae.yaml:
```diff
@@ -0,0 +1,156 @@
+    (:class:`~qiskit.circuit.classical.expr.Var`) used as the loop parameter of
+    :class:`~qiskit.circuit.ForLoopOp` instructions. A new ``LOOP_VARIABLE``
+    instruction-parameter type (type key ``w``) carries the variable declaration
+    in the ForLoop parameter slot using the existing ``EXPR_VAR_DECLARATION`` wire
```
```suggestion
    in the for-loop parameter slot using the existing ``EXPR_VAR_DECLARATION`` wire
```

maybe add a sentence about type key ``r``?

1ucian0 commented on releasenotes/notes/expr_range_feature-1091deab8f458eae.yaml:
```diff
@@ -0,0 +1,156 @@
+
+upgrade_qasm:
+  - |
+    Creating a :class:`~qiskit.qasm3.ast.ForLoopStatement` object without
```
This object doc is not indexed. I would put it more in terms of the openqasm code input than on internal api terms.

gadial commented on crates/circuit/src/classical/expr/expr.rs:
```diff
@@ -424,6 +531,33 @@ impl PyExpr {
         visitor.call_method1(intern!(visitor.py(), "visit_generic"), (slf,))
     }
+
+    /// Return a copy of this expression with classical :class:`Var` nodes substituted.
```
I'm not sure I understood the need for substitutions as part of this PR. Perhaps the whole substitution code can be factored to another PR? Even if this PR depends upon it, perhaps we can build this PR on top of it.

arthurostrauss replied:
The requirement came mainly for the immediate support of UnrollForLoops with loop parameter being a Var and for a const expr.Range: the idea is to create different circuits which, similarly to the parameter assignment case, carry different `expr.Value` that are binded to the circuit when the Range is contant. The substitution logic was implemented at the circuit level object, as well as additional methods in the control flow, to make the substitution recursive across all constructs for this pass to work (upon previous request by @1ucian0) . I agree that this is a pretty big feature for supporting something rather small (the only transpiler pass for ForLoops) as an existing Qiskit feature, but this could facilitate further developments for runtime expressions.

gadial commented on crates/circuit/src/classical/expr/range.rs:
```diff
@@ -0,0 +1,600 @@
+    pub stop: Expr,
+    pub step: Expr, // Always Some(Expr) - defaults to 1 when not provided
+    pub ty: Type,
+    pub constant: bool,
```
Why is this field needed?

arthurostrauss replied:
It indicates whether start, stop, and step are all compile-time-known expressions (no Vars). That’s what lets passes like `UnrollForLoops` decide whether a `ForLoopOp` with an `expr.Range` indexset can be statically unrolled, and what `Range.values()` relies on before materializing bounds.

Semantically it’s not independent state as it’s always start.const && stop.const && step.const, and we already recompute it in substitute and QPY load. It’s stored on the Rust struct for consistency with other expr nodes (Binary, Cast, etc.) and O(1) is_const() checks. It could equally be a derived getter on Range; Python already exposes it as read-only .const inferred at construction.

gadial commented on crates/circuit/src/classical/expr/range.rs:
```diff
@@ -0,0 +1,600 @@
+impl Range {
+    /// Returns the length of the range.
+    ///
+    /// Note: Returns 1 as the length is generally undefined for dynamic ranges.
```
This seems strange to me. Are we following a formal specification here?

arthurostrauss replied:
Range.len() isn’t specified by OpenQASM or any external IR doc. It exists so expr::Range can implement the same ForCollection::len() interface as PyRange/List. Because dynamic bounds have no static trip count, we return 1 (and is_empty() == false) as a conservative placeholder for recursive DAG size/depth and VF2 layout, which multiply by collection.len(). This is the minimal discrepancy I added (which is indeed strange for real time Ranges) to avoid breaking the existing other pieces mentioned. To make it less strange, I will modify this for the constant case to return the real length as if it was a Python range.

gadial commented on crates/circuit/src/classical/expr/range.rs:
```diff
@@ -0,0 +1,600 @@
+    ///
+    /// Note: Returns false if the object exists, as the emptiness cannot be
+    /// determined statically for dynamic ranges.
+    pub fn is_empty(&self) -> bool {
```
Again, something feels amiss here. We do we have to do it this way?

arthurostrauss replied:
Same reason as for len(), this is added to avoid dropping the expression in VF2 layout and in depth() for the dag.rs specification.

gadial commented on crates/circuit/src/classical/expr/range.rs:
```diff
@@ -0,0 +1,600 @@
+    /// Materialize a constant range as a Python ``range`` object.
+    ///
+    /// Uses Python ``range`` semantics (exclusive ``stop``).
+    pub fn values<'py>(&self, py: Python<'py>) -> PyResult<Bound<'py, PyRange>> {
```
I think the python facing method should be implemented in `PyRangeExpr`.

arthurostrauss replied:
Good point on layering. I've refactored so the split is explicit:
Range (language-agnostic IR) owns constness and materialization:

try_materialize_bounds() -> Result<(isize, isize, isize), MaterializeError> extracts static start/stop/step from the expression tree (exclusive-stop semantics, same as Python range).
len() / is_empty() already use this for constant ranges.

values() calls try_materialize_bounds() and then constructs a builtins.range via PyO3.
So I agree that allocating a Python object doesn't belong on Range as it shouldn't take a Python<'py> token or return Bound<'py, PyRange>. But the underlying operation isn't Python-specific: it's about whether the IR is constant and can be evaluated at compile time. That's the same information a C API would need (e.g. to expose static loop bounds for unrolling or validation), and it aligns with how we already represent literal ranges in ForCollection via the Rust operations::PyRange struct.

In short: materialization is an IR concern (Range); wrapping it as a Python range is a binding concern (PyRangeExpr). The public Python API (Range.values()) is unchanged, only where the PyO3 allocation lives moved.

gadial commented on crates/circuit/src/classical/expr/range.rs:
```diff
@@ -0,0 +1,600 @@
+                .into(),
+                _ => {
+                    return Err(PyErr::new::<pyo3::exceptions::PyTypeError, _>(
+                        "Range type must be an unsigned integer type",
```
This error is thrown if 1) the type of the range is not `Uint` **and** 2) the step is not given. This means that if I have a range with float values start=1, stop=2 and step=0.3, no error will be thrown? I don't think that's the desirable behavior.

Moreover, if we indeed rule out non-integer types, I see no reason not to implement `len` and `is_empty`. We can compute them on the fly for dynamic integers.

arthurostrauss replied:
Good catch. That error in the default-step match was actually redundant.

For a case like start=Uint(1), stop=Uint(2), step=Float(0.3): when step is provided, it is validated in py_value_to_expr, which rejects any Expr that is not Uint (so the float step raises "Range values must be of unsigned integer type" before we reach default-step logic).

When step is omitted, we synthesize step = 1 as a Uint in target_ty. That is safe because earlier validation already ensures start/stop are Uint, explicit ty= must be Uint, and the remaining check after type inference rejects any non-Uint target_ty. The extra _ arm in the default-step branch could never be hit in practice, so I removed it and kept the single post-inference Uint guard.

For your second point, this I think is not as trivial. When the Range is defined with real time Var (which is the intended general case, not only expr.Value), then we have no way at compile time of knowing how this should be unrolled. This use case arises when the Var comes from an input of the QuantumCircuit. Computing len and is_empty cannot be determined in those cases and that is why the defensive 1 and False are retained to avoid breaking the rest of the API for V2FLayout and dag.rs.

gadial commented on crates/circuit/src/classical/expr/range.rs:
```diff
@@ -0,0 +1,600 @@
+    }
+
+    fn __str__(&self, py: Python) -> PyResult<String> {
+        // Use get_start and get_stop methods for consistency
```
This seems overly complicated - why go through python conversions instead of grabbing the data directly from the rust fields?

arthurostrauss replied:
Fair point that this looks roundabout. The intent is to reuse each child’s canonical __repr__ rather than duplicating formatting logic in Rust: start/stop/step are Expr trees (Value, Var, Cast, …), and those types already define their Python __repr__. Converting to the Python wrapper and calling repr() keeps Range(...) consistent with nested nodes and matches how Binary, Cast, and Index do it elsewhere in the expr bindings.

gadial commented on crates/circuit/src/operations.rs:
```diff
@@ -482,18 +482,39 @@ pub enum ForCollection {
+/// `Symbol` / `expr::Var` types only provide `IntoPyObject` for owned values, not for
+/// references. All conversions to Python use owned `LoopVar` via clone.
+#[derive(Clone, Debug, IntoPyObject, FromPyObject, PartialEq, Eq)]
+pub enum LoopVar {
```
I'm using a similar mechanism in [this pr](https://github.com/Qiskit/qiskit/pull/16273) which might be merged in 2.5, so it'll take some adjustments.

arthurostrauss committed:
feat: materialize constant expr.Range bounds and match Python range semantics; add QPY v18 support

For constant expr.Range instances, materialize start/stop/step at compile time and compute accurate Python-like semantics for len() and emptiness. Key API and behaviour changes:

- Range.try_materialize_bounds is now public and returns (start, stop, step) for constant ranges; MaterializeError is public.
- Range.len() and Range.is_empty() use Python range semantics for constant ranges (falls back to conservative placeholder len == 1 / not empty for dynamic ranges).
- Introduced python_range_len(...) helper to compute Python's range length for given bounds.
- Removed the old Range::values that returned a PyRange; callers that need a Python range should use PyRangeExpr::values or try_materialize_bounds and construct the PyRange themselves.
- PyRangeExpr::values now materializes via try_materialize_bounds and returns a Python range object.
- Minor change in PyRangeExpr construction: default step value creation streamlined (still Uint-only).
- Added unit tests covering constant and non-constant len/is_empty behaviour.
- Documentation and QPY changelog updated for format version 18 to document RANGE expression and LOOP_VARIABLE support.

Breaking changes:
- Range::values removed from the public API; use Range::try_materialize_bounds (language-agnostic) or PyRangeExpr::values (produces a Python range) instead.

This makes compile-time reasoning about constant ranges precise (matching Python) and adds serialization/docs for the new QPY format elements.

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Addressing last writing suggestions

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Linting

arthurostrauss committed:
Merged main into branch: Updated LoopVar to LoopParam

Conform expr.Range / UnrollForLoops work to upstream #16273's input-var
loop counter (re-inferred in QPY), drop the branch's 'w' LOOP_VARIABLE
element, relax validation so range+Var is allowed, and fix the pyo3-0.29
subclass-init deprecation. Known limitation (Var-counter body cannot
capture outer vars) deferred to a follow-up PR.

arthurostrauss committed:
Preserve ForLoopOp loop parameter during substitute_vars and re-infer Var loop param on QPY read

When substituting variables in a ForLoopOp, exclude the loop's parameter (the body's input
variable) from substitutions so the body's input var isn't removed or mismatched with the
ForLoopOp.loop_parameter. This prevents accidental capture/removal of the loop counter during
substitution.

Also adjust QPY reading to handle ForLoopOp entries where the loop-parameter slot is serialized
as Null for runtime expr.Var counters: if the body circuit has exactly one input variable, treat
that as the loop parameter (matching the data model and the Rust reader behavior).

Update tests and release notes to reflect the loop-parameter semantics (Var loop counters are the
body's input variable and are re-inferred on QPY read). No public API signature changes, but this
behavior change affects substitution semantics and QPY deserialization for ForLoopOp.

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Use signed integers for ForCollection::List (Vec) to support negative/dynamic ranges

Change the ForCollection::List variant from Vec to Vec so list-based loop collections can represent negative indices or other signed integer ranges. This enables expressing dynamic/range-like collections that include negative values.

Breaking change: the ForCollection::List type in the public API now contains Vec instead of Vec; callers and pattern matches must be updated accordingly.

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

1ucian0 committed:
Merge branch 'main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
fix: roundtrip qpy

Clean up test/python/qpy/test_roundtrip.py by removing duplicated manual
write/read sequences and related assertions that duplicated what
assert_roundtrip_equal already verifies. Also remove an unused QpyError
import and correct a docstring to more precisely describe the limitation
on dumping expr.Range for-loops below QPY v18.

This is a test-only change (no runtime/API behavior change). It reduces
noise and duplication in the tests and makes expectations rely on the
central assert_roundtrip_equal helper.

arthurostrauss committed:
Linting

arthurostrauss committed:
additional linting

arthurostrauss committed:
Merge branch 'main' into dynamic_range_expr

arthurostrauss committed:
test(qpy): add helper to assert dump() rejects unsupported features and consolidate roundtrip tests

Introduce assert_dump_raises_unsupported to centralize checks that dumping a circuit with a given writer/version raises UnsupportedFeatureForVersion. Replace duplicated manual write_circuit/io.BytesIO blocks with calls to this helper across QPY roundtrip tests, and adjust writer iteration logic for versions that support the Rust writer.

This is a test-only refactor that reduces duplication and improves clarity; it does not change runtime or public API behavior.

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

gadial reviewed (dismissed):
I think this PR can be merged as part of the QPY18 update. Note the changes in the python QPY component are not required as python QPY does not support QPY18, but you can leave them.

gadial commented on qiskit/qasm3/exporter.py:
```diff
@@ -1226,22 +1241,34 @@ def build_for_loop(self, instruction: CircuitInstruction) -> ast.ForLoopStatemen
+            elif isinstance(indexset, expr.Range):
+                indexset_ast = ast.Range(
+                    start=self.build_expression(indexset.start),
+                    end=self.build_expression(indexset.stop),
```
Note that in the [QASM3 specification](https://openqasm.com/versions/3.0/language/classical.html), `As in other range expressions, the range is inclusive at both ends.` In Python ranges are not inclusive the end, so you may have to do `end=self.build_expression(indexset.stop+1)` here.

arthurostrauss replied:
Expr.Range intended to actually conform to OpenQASM3 specification by default, and not Python range style. This is the idea that everything that pertains the expr module should more or less be a direct reference to what gets eventually lowered to OpenQASM3. One who uses such a module would be expected to be aware of its lowering to the open qasm standard, as opposed to someone purely dealing with writing Python level ranges and who use ForLoop for wrapping up a simulator instruction.

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into dynamic_range_expr
